### PR TITLE
Avoid copying buffer from ReadOnlySequence as  ReceivedMqttPacket body

### PR DESCRIPTION
**Pull request by xljiulang**
### Related project
MQTTnet.AspNetCore

### How it affects
Read `ArraySegment<byte>` buffer from ReadOnlySequence as ReceivedMqttPacket body. 
It can greatly reduce unnecessary buffer-copying.

### Dependency
This request depend on #1592，after we make MqttBufferReader correctly support `ArraySegment<byte>`

### CODE REVIEW

**Comment by xljiulang** [comment]
Also depend on #1593

File: Source/MQTTnet.AspnetCore/ReaderExtensions.cs [hunk 1]
```diff
@@ -81,23 +83,24 @@ static bool TryReadBodyLength(ref ReadOnlySequence<byte> input, out int headerLe
             var index = 1;
             headerLength = 0;
             bodyLength = 0;
-            
-            var temp = input.Slice(0, Math.Min(5, input.Length)).GetMemory().Span;
+
+            var valueSequence = input.Slice(0, Math.Min(5, input.Length));
+            var valueSpan = valueSequence.IsSingleSegment ? valueSequence.First.Span : valueSequence.ToArray();
```
Comment (gfoidl) [inline Source/MQTTnet.AspnetCore/ReaderExtensions.cs:88]:
```suggestion
            var valueSpan = valueSequence.IsSingleSegment ? valueSequence.FirstSpan : valueSequence.ToArray();
```

Comment (gfoidl) [inline Source/MQTTnet.AspnetCore/ReaderExtensions.cs:88]: `valueSequence.Length == 5` so rather short, and lets assume that `valueSequence.IsSingleSegment == true / false` is both likely (though == true will be more likely) so `ToArray` will be expensive.

```c#
Span<byte> valueSpan = stackalloc byte[8];    // 8 as next power of 2
valueSequence.CopyTo(valueSpan);
valueSpan = valueSpan.Slice(0, valueSequence.Length);
```

`FirstSpan` is also not really cheap, so I'd just straighten the code and copy always. It's only a few bytes, but the code is much shorter so it should outweight the cost of copying.

File: Source/MQTTnet.AspnetCore/ReaderExtensions.cs [hunk 2]
```diff
@@ -81,23 +83,24 @@ static bool TryReadBodyLength(ref ReadOnlySequence<byte> input, out int headerLe
...
 
-                encodedByte = temp[index];
+                encodedByte = valueSpan[index];
                 index++;
 
                 value += (byte)(encodedByte & 127) * multiplier;
                 if (multiplier > 128 * 128 * 128)
                 {
-                    throw new MqttProtocolViolationException($"Remaining length is invalid (Data={string.Join(",", temp.Slice(1, index).ToArray())}).");
+                    throw new MqttProtocolViolationException($"Remaining length is invalid (Data={string.Join(",", valueSpan.Slice(1, index).ToArray())}).");
```
Comment (gfoidl) [inline Source/MQTTnet.AspnetCore/ReaderExtensions.cs:103]: This check should be moved down after L106 where `multiplier` is updated.
Then as this materializes a costly string it should use a throw-helper. Like
```c#
// L105
multiplier *= 128;

if (multiplier > 128 * 128 * 128)
{
    Throw(valueSpan, index);

    static void Throw(ReadOnlySpan<byte> valueSpan, int index)
        => throw new MqttProtocolViolationException($"Remaining length is invalid (Data={string.Join(",", valueSpan.Slice(1, index).ToArray())}).");
}
```
This keeps the loop-body short (from CPU-view), thus faster.

File: Source/MQTTnet.AspnetCore/ReaderExtensions.cs (hunk 1 above)
Comment (xljiulang) [reply Source/MQTTnet.AspnetCore/ReaderExtensions.cs:88]: Benefit a lot！

Comment (xljiulang) [reply Source/MQTTnet.AspnetCore/ReaderExtensions.cs:88]: We cannot use `FirstSpan` Api under some target frameworks, such as `netcoreapp` and `netstandard`

File: Source/MQTTnet.AspnetCore/ReaderExtensions.cs [hunk 3]
```diff
@@ -61,15 +62,16 @@ public static class ReaderExtensions
             return true;
         }
 
-        static ReadOnlyMemory<byte> GetMemory(this in ReadOnlySequence<byte> input)
+        static ArraySegment<byte> AsArraySegment(ref ReadOnlySequence<byte> input)
```
[the remark this answers is not in the document]
Comment (xljiulang) [reply Source/MQTTnet.AspnetCore/ReaderExtensions.cs:65]: There cannot be interpreted as `ReadOnlyMemory` - > `ArraySegment`, but `ReadOnlyMemory.ToArray()` - > `MemoryMarshal. TryGetArray(ReadOnlyMemory, out ReadOnlyMemory)`

File: Source/MQTTnet.AspnetCore/ReaderExtensions.cs (hunk 2 above)
Comment (xljiulang) [reply Source/MQTTnet.AspnetCore/ReaderExtensions.cs:103]: Great!!

File: Source/MQTTnet.AspnetCore/ReaderExtensions.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (gfoidl) [reply Source/MQTTnet.AspnetCore/ReaderExtensions.cs:65]: Yep, I just mean there should be a benchmark showing that it doesn't get worse. I'm uncertain in which way the needle will move, so just a measure of caution. I'm sure you have this on your table 😄.

[the remark this answers is not in the document]
Comment (xljiulang) [reply Source/MQTTnet.AspnetCore/ReaderExtensions.cs:65]: Well, I will provide a Benchmark as soon as possible.

**Comment by xljiulang** [comment]
The benchmark result shows a good reduction in memory allocation!
``` ini

BenchmarkDotNet=v0.13.2, OS=Windows 10 (10.0.19044.2251/21H2/November2021Update)
Intel Core i3-4150 CPU 3.50GHz (Haswell), 1 CPU, 4 logical and 2 physical cores
.NET SDK=7.0.100
  [Host]   : .NET 6.0.8 (6.0.822.36306), X64 RyuJIT AVX2
  .NET 6.0 : .NET 6.0.8 (6.0.822.36306), X64 RyuJIT AVX2

Job=.NET 6.0  Runtime=.NET 6.0  

```
| Method |     Mean |     Error |    StdDev | Ratio | Rank |    Gen0 | Allocated | Alloc Ratio |
|------- |---------:|----------:|----------:|------:|-----:|--------:|----------:|------------:|
| Before | 3.593 μs | 0.0124 μs | 0.0097 μs |  1.00 |    2 | 13.4201 |  20.57 KB |        1.00 |
|  After | 2.536 μs | 0.0151 μs | 0.0134 μs |  0.71 |    1 |  6.8932 |  10.54 KB |        0.51 |

**Pull request merged**